src/main/java/org/jahia/modules/appshell/Main.java

```diff
@@ -144,49 +218,119 @@ private void setCustomAttributes(JahiaUser currentUser, HttpServletRequestWrappe
         wrapper.setAttribute("config", config.toString());
     }
 
+    public void updateBundleInfos() {
+        List<Bundle> packages = getPackages();
+
+        List<Pattern> newUrlPatterns = new ArrayList<>();
+        Map<Bundle, JSONObject> newBundleDescriptors = new LinkedHashMap<>();
```

Comment:
I'm a worried about concurrency, what happen in case requests are incoming in between a bundle changed event and the bundle infos are not yet recalculated ?
I think it could happen that in rare race condition request will face empty:
```
 List<Pattern> newUrlPatterns = new ArrayList<>();
        Map<Bundle, JSONObject> newBundleDescriptors = new LinkedHashMap<>();
```
Leading to error in the UI due to incomplete JS scripts not loaded.

Reply:
I deliberately use new structures that are atomically swapped at the end of the method to avoid changing things concurrently. So it is still possible that a race condition happens at the swap but that should be extremely unlikely. I use this method to avoid introducing locks and keep good performance while minimizing the risk of race conditions.